### PR TITLE
Allow easing testing of java ea versions using gradle java tool chain support

### PR DESCRIPTION
**Pull request by breskeby**
This introduces the ability to simply configure a java tool chain for elasticsearch java projects to be used.
If an environment variable `JAVA_TOOLCHAIN_HOME `is declared, this JDK will be used as toolChain in elasticsearch.java projects.

This should make testing our build with java `ea` versions easier and allows detangling the used compiler jdk from the gradle java runtime.

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java [hunk 1]
```diff
@@ -106,11 +106,10 @@ private static void disableTransitiveDependenciesForSourceSet(Project project, S
      */
     public static void configureCompile(Project project) {
         project.getExtensions().getExtraProperties().set("compactProfile", "full");
-
         JavaPluginExtension java = project.getExtensions().getByType(JavaPluginExtension.class);
+        BuildParams.getJavaToolChainSpec().map(spec -> java.toolchain(spec)).getOrNull();
```
Comment (mark-vieira) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java:110]: Having side-effects in a `map()` call seems a bit like bad practice. I think being explicit here and just putting this behind a call to `ifPresent()` seems more idiomatic.

**Review by mark-vieira — COMMENTED** [review]
Question: Is this toolchain used for both compilation and test execution? If so, is that the behavior we want?

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java (hunk 1 above)
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java]: I wish gradle would have a more functional api here. Changed.

**Comment by breskeby** [comment]
@mark-vieira yes declaring a tool chain at the java extension level actually applies that toolchain to all compile, test and javadoc tasks. We could split this up and only apply it to compile tasks if required. I think having all set at once seems the reasonable default. We can put more effort into this and split this up or only apply it to compile tasks. What would you think is a most useful default here @ChrisHegarty @rjernst?

**Comment by rjernst** [comment]
> What would you think is a most useful default here

I think apply to compile, test and javadoc is a reasonable default.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java (hunk 1 above)
Comment (mark-vieira) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/ElasticsearchJavaBasePlugin.java]: Agreed. Might be a good simple submission. An `ifPresent()` method like exists on `java.util.Optional` that accepts a `Consumer` would be super useful.

**Comment by mark-vieira** [comment]
If that's the case then it could be confusing if combined with `RUNTIME_JAVA_HOME`. We should make that explicit and enforce mutual exclusivity or at least warn the user that `RUNTIME_JAVA_HOME` will be ignored.

**Comment by rjernst** [comment]
> We should make that explicit and enforce mutual exclusivity or at least warn the user that RUNTIME_JAVA_HOME will be ignored.

I don't think it needs to be a warning or ignored. Most of the time RUNTIME_JAVA_HOME is not set. Can we make it so that RUNTIME_JAVA_HOME is still used for tests *if* it is specified?

**Comment by mark-vieira** [comment]
> Can we make it so that RUNTIME_JAVA_HOME is still used for tests _if_ it is specified?

This would be a good option if possible but I'm not sure the toolchain apis make that distinction.

**Comment by breskeby** [comment]
From looking at the code using RUNTIME_JAVA_HOME for tests seems like a breaking change. 

Currently we only configure runtime java home for the test cluster runtime and this is not affected by this change here.

What I changed here is that when resolving RUNTIME_JAVA I fallback to toolchain java version if set and if not we fall back as usual to the jdk used by the gradle runtime (as we did before).

**Review by rjernst — APPROVED** [review]
I think the behavior here is what I would expect. Runtime Java home is used if it is found, otherwise fallback to the tool chain. Thanks Rene.

**Review by ChrisHegarty — APPROVED** [review]
LGTM.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/info/GlobalBuildInfoPlugin.java [hunk 2]
```diff
@@ -164,6 +181,10 @@ private void logGlobalBuildInfo() {
             LOGGER.quiet("  JDK Version           : " + gradleJvmImplementationVersion + " (" + gradleJvmVendorDetails + ")");
             LOGGER.quiet("  JAVA_HOME             : " + gradleJvm.getJavaHome());
         }
+        String javaToolchainHome = System.getenv("JAVA_TOOLCHAIN_HOME");
+        if (javaToolchainHome != null) {
+            LOGGER.quiet("  JAVA TOOLCHAIN HOME   : " + javaToolchainHome);
```
[the remark this answers is not in the document]
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/info/GlobalBuildInfoPlugin.java]: 👍 fixed

**Pull request merged**